### ISSUE TITLE
cut missing in xinteger//1

### ISSUE THREAD

**Issue by jburse** [SWI-Prolog/swipl-devel#475]
Hi,

Was playing around with library(dcg/basics). I got the following crash:

```
?- phrase(xinteger(123), L).
L = [55, 98] ;
ERROR: Stack limit (1.0Gb) exceeded
```

Possibly a missing cut. When I compare integer//1 and xinteger//1 I see a difference:

```
integer(I, Head, Tail) :-
	nonvar(I), !,
	format(codes(Head, Tail), '~d', [I]).

xinteger(Val, Head, Tail) :-
	integer(Val),
	format(codes(Head, Tail), '~16r', [Val]).
```
The stack overflow doesn't happen with integer//1.

**Comment by jburse** [comment]
A further, somehow related issue, xinteger//1 unparsing can do negative numbers:

```
?- phrase(xinteger(-123), L).
L = [45, 55, 98] .
```

But parsing cannot take negative numbers:

```
?- phrase(xinteger(X), [45, 55, 98]).
false.
```

**Comment by JanWielemaker** [comment]
Thanks.  Fixed with cbf29b43ef47a48864776c6f4b309f028f3f6658
